### PR TITLE
SQlite 3.19.3 and Socket::select fix

### PR DESCRIPTION
**Pull request by micheleselea**
changelog
http://www.sqlite.org/releaselog/3_19_3.html

### CODE REVIEW

**Comment by micheleselea** [comment]
Working with this SQlite version since June, and no problem in production environment

**Comment by micheleselea** [comment]
about epoll_event memset fix: with -std=c++14 sizeof(eventsOut) gives wrong result, so memset fail. This issue affect only Linux compiled in c++14 with epoll.
The function is used only by socketreactor, that can cause a while(1) in the thread that does select request

**Comment by aleks-f** [comment]
it fails to compile on 4.8:
```
[0m/home/travis/build/pocoproject/poco/Net/src/Socket.cpp: In static member function â€˜static int Poco::Net::Socket::select(Poco::Net::Socket::SocketList&, Poco::Net::Socket::SocketList&, Poco::Net::Socket::SocketList&, const Poco::Timespan&)â€™:
/home/travis/build/pocoproject/poco/Net/src/Socket.cpp:83:48: error: variable-sized object â€˜eventsInâ€™ may not be initialized
   struct epoll_event eventsIn[epollSize] = { 0 };  
                                                ^
/home/travis/build/pocoproject/poco/Net/src/Socket.cpp:168:48: error: variable-sized object â€˜eventsOutâ€™ may not be initialized
  struct epoll_event eventsOut[epollSize] = { 0 }; 
                                                ^
make[2]: *** [Net/CMakeFiles/Net.dir/src/Socket.cpp.o] Error 1
make[1]: *** [Net/CMakeFiles/Net.dir/all] Error 2
```
I am surprised that it fails to calculate sizeof on c++14 and that the fix compiles at all because it's a dynamic sized array.

Anyway, could this be a fix:
```
struct epoll_event eventsIn[epollSize];
memset(eventsIn, 0, sizeof(epoll_event) * epollSize );
```

**Comment by micheleselea** [comment]
Yes I'll try to put a check on c++xx version used...

I manage to remove the pull request



*Michele Pradella*
/R&D Software Engineer
/ michele.pradella@selea.com <mailto:michele.pradella@selea.com>
Office: +39 0375 889091

------------------------------------------------------------------------
	
------------------------------------------------------------------------
Selea s.r.l

V. Aldo Moro, 69
46019 Cicognara MN - Italy
Phone: +390375889091
Fax: +390375889080
http://www.selea.com

	

Note: The information contained in this message may be privileged and 
confidential and protected from disclosure. If the reader of this 
message is not the intended recipient, or an employee or agent 
responsible for delivering this message to the intended recipient, you 
are hereby notified that any dissemination, distribution or copying of 
this communication is strictly prohibited. If you have received this 
communication in error, please notify us immediately by replying to the 
message and deleting it from your computer. Thank you.

Il 25/07/2017 11.19, Aleksandar Fabijanic ha scritto:

**Comment by micheleselea** [comment]
If you try you will see that sizeof(eventsIn)will give you bytes count 
with c++11 and element number in c++14

Someone find the same problem
http://wm.ite.pl/notesen/2015-03-22-compiler-warnings.html

*Michele Pradella*
/R&D Software Engineer
/ michele.pradella@selea.com <mailto:michele.pradella@selea.com>
Office: +39 0375 889091

------------------------------------------------------------------------
	
------------------------------------------------------------------------
Selea s.r.l

V. Aldo Moro, 69
46019 Cicognara MN - Italy
Phone: +390375889091
Fax: +390375889080
http://www.selea.com

	

Note: The information contained in this message may be privileged and 
confidential and protected from disclosure. If the reader of this 
message is not the intended recipient, or an employee or agent 
responsible for delivering this message to the intended recipient, you 
are hereby notified that any dissemination, distribution or copying of 
this communication is strictly prohibited. If you have received this 
communication in error, please notify us immediately by replying to the 
message and deleting it from your computer. Thank you.

Il 25/07/2017 11.19, Aleksandar Fabijanic ha scritto:

**Comment by micheleselea** [comment]
@aleks-f  can I need to know a thing about pullrequest. If I have more changes to commit, should I continue commit in my current fork and create another pullrequest? or should I re fork it?

**Comment by aleks-f** [comment]
this one is merged, create another fork from develop branch and send new pull

**Comment by aleks-f** [comment]
generally speaking it is best to keep pulls focused on one specific thing (also create an issue if it's a bug so we can track it)

**Comment by micheleselea** [comment]
ok thanks, sorry for this question but was the first time of pull request and I want to make it in the correct way. I suppose that the correct way was to split the modification one for "kind" (one for SQLite, and one for socket) but I misunderstood the good way to do it.
Tanks for informations

**Comment by zosrothko** [comment]
@micheleselea I was asking to comment the Socket.cpp code itself, not this pull request! ;)

**Comment by micheleselea** [comment]
ah ok understand I'll do it as soon as I can.  

sorride but was my first pullrequest since long time  

Il f2017-08-02 08:27 zosrothko ha scritto:

> @micheleselea [1] I was asking to comment the Socket.cpp code itself, not this pull request! ;) 
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [2], or mute the thread [3].

 

Links:
------
[1] https://github.com/micheleselea
[2] https://github.com/pocoproject/poco/pull/1825#issuecomment-319580139
[3]

File: Net/src/Socket.cpp [hunk 1]
```diff
@@ -79,9 +79,10 @@ int Socket::select(SocketList& readList, SocketList& writeList, SocketList& exce
 	if (epollSize == 0) return 0;
 
 	int epollfd = -1;
-	{
+	{		
 		struct epoll_event eventsIn[epollSize];
-		memset(eventsIn, 0, sizeof(eventsIn));
+		memset(eventsIn, 0, sizeof(epoll_event) * epollSize );
```
Comment (micheleselea) [inline Net/src/Socket.cpp]: gcc with -std=c++11 sizeof(eventsIn) gives bytes number
gcc with -std=c++14 sizeof(eventsIn) gives elements number
changed memset to keep it save in both condition

File: Net/src/Socket.cpp [hunk 2]
```diff
@@ -167,7 +168,7 @@ int Socket::select(SocketList& readList, SocketList& writeList, SocketList& exce
 	}
 
 	struct epoll_event eventsOut[epollSize];
-	memset(eventsOut, 0, sizeof(eventsOut));
+	memset(eventsOut, 0, sizeof(epoll_event) * epollSize );
```
Comment (micheleselea) [inline Net/src/Socket.cpp]: gcc with -std=c++11 sizeof(eventsOut) gives bytes number
gcc with -std=c++14 sizeof(eventsOut) gives elements number
changed memset to keep it safe

**Comment by aleks-f** [comment]
@micheleselea : the array size fix does not hurt, but it must be a g++ version oddity/bug or something like that because I can not reproduce it, eg.:

```
$ clang++ --version
Apple LLVM version 8.1.0 (clang-802.0.42)

$ cat array.cpp
#include <iostream>

int main()
{
    int sz = 10;
    int a[10], b[sz];
    std::cout << "size of array of 10 int: " << sizeof(int[10])   << '\n'
              << "size of array a of 10 int: " << sizeof a          << '\n'
              << "size of array b of 10 int: " << sizeof b << '\n';
}

$ clang++ -std=c++14 -o array14 array.cpp
$ clang++ -std=c++11 -o array11 array.cpp

$ ./array11
size of array of 10 int: 40
size of array a of 10 int: 40
size of array b of 10 int: 40

$ ./array14
size of array of 10 int: 40
size of array a of 10 int: 40
size of array b of 10 int: 40
```

```
$ g++ --version
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609

$ g++ -std=c++14 -o array14 array.cpp
$ g++ -std=c++11 -o array11 array.cpp

$ ./array11
size of array of 10 int: 40
size of array a of 10 int: 40
size of array b of 10 int: 40

$ ./array14
size of array of 10 int: 40
size of array a of 10 int: 40
size of array b of 10 int: 40
```

**Comment by micheleselea** [comment]
I don't know I was able to reproduce on a Debian jessie (3.16.0-4amd64) with gcc version 4.9.2-10 and on a Raspberry with DietPi 4.1.17-v7+ distribution and gcc version 4.9.2-10
I use -std=c++11 or 14 in the config.make file with POCO_CFLAGS (I'm not in office so I'm try remember best as I can).
Could be a problem of that gcc version?

**Comment by zosrothko** [comment]
@aleks-f Your test is not a good one, neither the compiler. Below a relevant test
```
fandre@ubuntu-trusty-64:~/gcc$ cat array.cpp
#include <iostream>
using namespace std;

int size(int n) {
    char array[n];
    return sizeof(array);
}

int main() {
        cout << "size(5)=" << size(5) <<endl;
        return 0;
}
fandre@ubuntu-trusty-64:~/gcc$ g++-4.9 -std=c++11 array.cpp -o array11
fandre@ubuntu-trusty-64:~/gcc$ g++-4.9 -std=c++14 array.cpp -o array14
array.cpp: In function ‘int size(int)’:
array.cpp:6:24: warning: taking sizeof array of runtime bound [-Wvla]
     return sizeof(array);
                        ^
fandre@ubuntu-trusty-64:~/gcc$ ./array11
size(5)=5
fandre@ubuntu-trusty-64:~/gcc$ ./array14
size(5)=1
fandre@ubuntu-trusty-64:~/gcc$
```

**Comment by zosrothko** [comment]
@micheleselea Can you reference this [url ](http://wm.ite.pl/notesen/2015-03-22-compiler-warnings.html) in your comment; Thanks

**Comment by aleks-f** [comment]
@zosrothko: yes and no - I understand why compiler warns in your example - it can't see in the function that in this particular case it's dealing with a constant known at compile time. but, what is confusing to me is that a compiler would return number of elements as sizeof and array - that sounds like something that would break lots of code.

**Comment by zosrothko** [comment]
@aleks-f Right and as this [paper](http://wm.ite.pl/notesen/2015-03-22-compiler-warnings.html) states, this feature has been removed from the standard, but it is kept in g++-4.9. Thus one have to deal with it since Poco should support g++ 4.9.
```
Weird? Yes and no. GCC implemented proposal N3639 for C++14 ("Runtime-sized arrays with automatic storage duration") where sizeof(VLA) is an error ("ill-formed"), but as we see authors have chosen a soft warning rather a compile error. BTW, N3639 was rejected and GCC also removed support for this proposal.
```

**Comment by micheleselea** [comment]
@zosrothko ok I'll put in the code as soon I came back from holidays.
@aleks-f I think that the point here is not when or when not the size of give correct or wrong results. Probably, generally speaking, is not a good idea do sizeof(vector pointer) because the behavior could be unpredictable. I mean for example that the compiler could give you the size of the pointer (just an example) instead what we need here in memset. Doing it explicitly I think it's the best way to reach the result without need to consider compilers versions or other variables.
I will never use sizeof that way because for me it's not clear. Anyway I think the solution proposed should solve ambiguity anyway

**Pull request merged**